git_test.go

```diff
@@ -4,72 +4,198 @@ import (
 
-		srv := httptest.NewServer(gittest.NewServer(serverFS))
+				cloned, err := envbuilder.CloneRepo(context.Background(), envbuilder.CloneRepoOptions{
+					Path:    "/workspace",
+					RepoURL: srvURL,
+					Storage: clientFS,
+					RepoAuth: &githttp.BasicAuth{
+						Username: tc.username,
+						Password: tc.password,
+					},
+				})
+				require.Equal(t, tc.expectClone, cloned)
+				if tc.expectError != "" {
+					require.ErrorContains(t, err, tc.expectError)
+					return
+				}
+				require.NoError(t, err)
+				require.True(t, cloned)
 
-		clientFS := memfs.New()
-		cloned, err := envbuilder.CloneRepo(context.Background(), envbuilder.CloneRepoOptions{
-			Path:    "/workspace",
-			RepoURL: srv.URL,
-			Storage: clientFS,
-		})
-		require.NoError(t, err)
-		require.True(t, cloned)
+				readme := mustRead(t, clientFS, "/workspace/README.md")
+				require.Equal(t, "Hello, world!", readme)
+				gitConfig := mustRead(t, clientFS, "/workspace/.git/config")
+				require.Contains(t, gitConfig, srvURL)
+			})
 
-		file, err := clientFS.OpenFile("/workspace/README.md", os.O_RDONLY, 0644)
-		require.NoError(t, err)
-		defer file.Close()
-		content, err := io.ReadAll(file)
-		require.NoError(t, err)
-		require.Equal(t, "Hello, world!", string(content))
-	})
+			// In-URL-style auth e.g. http://user:password@host:port
+			t.Run("InURL", func(t *testing.T) {
+				t.Parallel()
+				srvURL := setupGit(t, tc.srvUsername, tc.srvPassword)
+				authURL, err := url.Parse(srvURL)
+				require.NoError(t, err)
+				authURL.User = url.UserPassword(tc.username, tc.password)
+				clientFS := memfs.New()
 
-	t.Run("DoesntCloneIfRepoExists", func(t *testing.T) {
-		t.Parallel()
-		clientFS := memfs.New()
-		gittest.NewRepo(t, clientFS)
-		cloned, err := envbuilder.CloneRepo(context.Background(), envbuilder.CloneRepoOptions{
-			Path:    "/",
-			RepoURL: "https://example.com",
-			Storage: clientFS,
+				cloned, err := envbuilder.CloneRepo(context.Background(), envbuilder.CloneRepoOptions{
+					Path:    "/workspace",
+					RepoURL: authURL.String(),
+					Storage: clientFS,
+				})
+				require.Equal(t, tc.expectClone, cloned)
+				if tc.expectError != "" {
+					require.ErrorContains(t, err, tc.expectError)
+					return
+				}
+				require.NoError(t, err)
+				require.True(t, cloned)
+
+				readme := mustRead(t, clientFS, "/workspace/README.md")
+				require.Equal(t, "Hello, world!", readme)
+				gitConfig := mustRead(t, clientFS, "/workspace/.git/config")
+				// We do not modify the git URL that folks pass in.
+				require.Contains(t, gitConfig, authURL.String())
+			})
 		})
-		require.NoError(t, err)
-		require.False(t, cloned)
+	}
+}
+
+func mustRead(t *testing.T, fs billy.Filesystem, path string) string {
+	t.Helper()
+	f, err := fs.OpenFile(path, os.O_RDONLY, 0644)
+	require.NoError(t, err)
+	content, err := io.ReadAll(f)
+	require.NoError(t, err)
+	return string(content)
+}
+
+func setupGit(t *testing.T, user, pass string) (url string) {
```

Comment:
What about moving this to `gittest`?

Reply:
I'll open a a separate PR to refactor this 👍